### PR TITLE
updated dqn.py after discussion

### PR DESCRIPTION
**Pull request by akshatgarg06**
**CORE Summary**

- Introduced a DQN-based attacker model registered under dqn_attacker.

- Align training/validation/test steps with existing Q‑model patterns (TD loss, action masking, metrics).

- Keep optimizer/scheduler wiring consistent with the current modular architecture.

### CODE REVIEW

**Review by kenjiro-mk — CHANGES_REQUESTED** [review]
### Follow-up on TD Target Calculation Issue
Thank you for your work. I have reviewed the code, but it appears that the issue regarding TD target calculation—which I pointed out in the previous review—has not yet been addressed.

Affected Locations:
- `training_step`: Lines 93–95
- `validation_step`: Lines 150–152
- `test_step`: Lines 228–230

Current Implementation:
In all three methods, `q_values` calculated via `forward()` are being reused inside the `torch.no_grad()` block.

Recommended Fix:
Please call `self.forward(batch)` again specifically within the `torch.no_grad()` block to ensure you obtain Q-values that are completely detached from the computational graph:

```
with torch.no_grad():
    # Re-calculating to ensure next_q_values are detached from the graph
    next_q_values = self.forward(batch)[:, 1:].max(dim=2).values
    td_target = reward[:, :-1] + self.gamma * next_q_values
```

This change will make the DQN training more stable and ensure the implementation is theoretically sound by preventing unintended gradient flow through the targets.

Please let me know once you have had a chance to verify this.

**Comment by akshatgarg06** [comment]
@kenjiro-mk  I was testing 2-3 approaches to implementing the recommended changes suggested for DQN.

It resulted in a slight delay in my execution.

**Review by kenjiro-mk — APPROVED** [review]
